### PR TITLE
Rule S4428: "PartCreationPolicyAttribute" should be used with "Export…

### LINKED ISSUE #1166: Rule S4428: "PartCreationPolicyAttribute" should be used with "ExportAttribute"
[RSPEC-4428](https://jira.sonarsource.com/browse/RSPEC-4428)

The PartCreationPolicyAttribute attribute, which is part of the Managed Extensibility Framework (MEF), is used to specify how the exported object will be created. Therefore it doesn't make sense not to export this a class with this attribute using the ExportAttribute attribute.

### PR DESCRIPTION
**Pull request by valhristov**
…Attribute"

Fix #1166

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var attribute = (AttributeSyntax)c.Node;
+
+                    if (IsPartCreationPolicyAttribute(attribute) &&
+                        attribute.FirstAncestorOrSelf<MemberDeclarationSyntax>() is ClassDeclarationSyntax classDeclaration)
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: PartCreationPolicy is not allowed on other than classes

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class PartCreationPolicyShouldBeUsedWithExportAttribute : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4428";
+        private const string MessageFormat = "Add the 'ExportAttribute' or remove 'PartCreationPolicyAttribute' to/from this class definition.";
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: [cosmetic][minor] Split the line so we can easily review on GitHub.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        private static readonly ISet<KnownType> ExportAttributes = new HashSet<KnownType>
+        {
+            KnownType.System_ComponentModel_Composition_ExportAttribute,
+            KnownType.System_ComponentModel_Composition_InheritedExportAttribute,
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: [cosmetic][minor] Move `c =>` to a new line.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                {
+                    var attribute = (AttributeSyntax)c.Node;
+
+                    if (IsPartCreationPolicyAttribute(attribute) &&
+                        attribute.FirstAncestorOrSelf<MemberDeclarationSyntax>() is ClassDeclarationSyntax classDeclaration)
+                    {
+                        var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                        if (classSymbol != null &&
+                            !classSymbol.GetAttributes().Any(IsExportAttribute))
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: I don't think this is the right check. AFAIK the goal of `InheritedExport` is to say you don't have to specify the `Export` on sub-classes so your rule would raise an issue where it shouldn't. You will have to walk up the tree looking for `InheritedExport`.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.ComponentModel.Composition;
+
+namespace Tests.Diagnostics
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: Please add tests with other `CreationPolicy` values.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var attribute = (AttributeSyntax)c.Node;
+
+                    if (IsPartCreationPolicyAttribute(attribute) &&
+                        attribute.FirstAncestorOrSelf<MemberDeclarationSyntax>() is ClassDeclarationSyntax classDeclaration)
+                    {
+                        var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+                        if (classSymbol == null ||
+                            classSymbol.GetAttributes().Any(IsExportAttribute) ||
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: Export is not inherited

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.cs (hunk 5 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: Isn't the CreationPolicy irrelevant?

Comment (Evangelink) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: It's irrelevant hence adding a test to show it is!

Comment (valhristov) [reply sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/PartCreationPolicyShouldBeUsedWithExportAttribute.cs]: There are many irrelevant things for this rule :)

**Pull request merged**